samples/MQTTSN/src/paho_iot_hub_telemetry_example.c

```diff
@@ -0,0 +1,669 @@
+      ENV_MQTTSN_GATEWAY_ADDRESS,
+      DEFAULT_GATEWAY_ADDRESS,
+      false,
+      gateway_address_span,
+      &gateway_address_span));
+
+  az_span gateway_port_span = AZ_SPAN_FROM_BUFFER(gateway_port);
+  AZ_RETURN_IF_FAILED(read_configuration_entry(
+      ENV_MQTTSN_GATEWAY_PORT,
+      ENV_MQTTSN_GATEWAY_PORT,
+      DEFAULT_GATEWAY_PORT,
+      false,
+      gateway_port_span,
+      &gateway_port_span));
+
+  // Initialize the hub client with the hub host endpoint and the default connection options
+  AZ_RETURN_IF_FAILED(az_iot_hub_client_init(
+      client,
+      az_span_slice(iot_hub_hostname_span, 0, (int32_t)strlen(iot_hub_hostname)),
+      az_span_slice(device_id_span, 0, (int32_t)strlen(device_id)),
+      NULL));
+
+  // Save gateway address to context
+  az_span_to_str(
+      gateway_address,
+      sizeof(gateway_address),
+      az_span_slice(gateway_address_span, 0, (int32_t)strlen(gateway_address)));
+  *host = gateway_address;
+
+  // Save gateway port to context
+  AZ_RETURN_IF_FAILED(
+      az_span_atou32(az_span_slice(gateway_port_span, 0, (int32_t)strlen(gateway_port)), port));
+
+  return AZ_OK;
+}
+
+/*
+ * Read the Environment Variables and and initialize the az_iot_hub_client
+ */
+static int init_client_context(iothub_client_context* ctx)
+{
+  int rc;
+  ctx->packet_id = 0;
+
+  if (az_failed(rc = read_configuration_and_init_client(&ctx->client, &ctx->host, &ctx->port)))
+  {
+    printf("Failed to read configuration from environment variables, return code %d\r\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+/*
+ * First 10 attempts: try within 3 seconds
+ * Next 10 attempts: retry after every 1 minute
+ * After 20 attempts: retry every 10 minutes
+ */
+int get_connection_timeout(int attempt_number)
+{
+  return (attempt_number < 10) ? 3 : (attempt_number < 20) ? 60 : 600;
+}
+
+/*
+ * 1. Create CONNECT packet
+ * 2. Send CONNECT packet to the MQTTSN Gateway
+ */
+static int send_connect(iothub_client_context* ctx, MQTTSNPacket_connectData* options)
+{
+  int rc;
+  int len;
+
+  // 1. Create CONNECT packet
+  if ((len = MQTTSNSerialize_connect(scratch_buffer, sizeof(scratch_buffer), options)) == 0)
+  {
+    printf("Failed to serialize CONNECT packet, return code %d\r\n", len);
+    return len;
+  }
+
+  // 2. Send CONNECT packet to the MQTTSN Gateway
+  if ((rc = transport_sendPacketBuffer(ctx->host, ctx->port, scratch_buffer, len)) != 0)
+  {
+    printf("Failed to send CONNECT packet to the Gateway, return code %d\r\n", rc);
+    return rc;
+  }
+
+  return 0;
+}
+
+/*
+ * Wait for CONNACK packet from the MQTTSN Gateway
+ */
+static int receive_connack()
+{
+  int rc;
+
+  if (MQTTSNPacket_read(scratch_buffer, sizeof(scratch_buffer), transport_getdata)
+      == MQTTSN_CONNACK)
+  {
+    if (MQTTSNDeserialize_connack(&rc, scratch_buffer, sizeof(scratch_buffer)) != 1 || rc != 0)
+    {
+      printf("Failed to deserialize CONNACK packet, return code %d\r\n", rc);
+    }

```

Comment:
if deserialization fails rc will be not set. Can you set rc = -1; after line 231